bpf/lib/encrypt.h

```diff
@@ -132,6 +133,84 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 	return ctx_redirect(ctx, CILIUM_IFINDEX, 0);
 #endif /* ENABLE_ROUTING */
 }
+
+#if defined(ENABLE_IPV4) && defined(TUNNEL_MODE)
+/* Sets the encryption mark on an overlay (VXLAN) packet and redirects the
+ * packet to the ingress side of it's associated ifindex.
+ *
+ * The recirculated overlay packet will then be subjected to XFRM hooks in the
+ * output routing path, since the original src/dst of the overlay packet routes
+ * off-host.
+ *
+ * This function is useful when you want to encrypt overlay traffic and use the
+ * underlay to deliver encrypted overlay traffic to the remote node.
+ * For this to work the IPSec control plane must install XFRM policies and
+ * states which set the tunnel source and destination to the underlay address of
+ * the destination node.
+ *
+ * If the redirect to the ingress side of ctx->ingress is successful
+ * CTX_ACT_REDIRECT is returned, otherwise an error code is returned.
+ *
+ * Be aware that the redirected-to interface needs to have the following
+ * sysctl enabled for this to work correctly (per-device is fine)
+ *   - net.ipv4.conf.default.rp_filter = 0
+ *   - net.ipv4.conf.default.accept_local = 1
+ */
+static __always_inline int
+encrypt_overlay_and_redirect(struct __ctx_buff *ctx, void *data,
+			     void *data_end, struct iphdr *ip4)
+{
+	struct endpoint_info *ep_info = NULL;
+	struct iphdr *inner_ipv4 = NULL;
+	__u8 encrypt_key = 0;
+	__u8 dst_mac = 0;
+	int ret = 0;
+
+	ret = tunnel_vxlan_get_inner_ipv4(data, data_end, ip4, &inner_ipv4);
+	if (!ret)
+		return DROP_INVALID;
+
+	ep_info = __lookup_ip4_endpoint(inner_ipv4->saddr);
+	if (!ep_info)
+		return DROP_INVALID;
+
+	/*
+	 * this is a vxlan packet so ip4->daddr is the tunnel endpoint
+	 */
+	ret = set_ipsec_encrypt(ctx, encrypt_key, ip4->daddr, ep_info->sec_id,
+				false);
+	if (ret != CTX_ACT_OK)
+		return ret;
+
+	/*
+	 * source mac is our current egress interface, lets copy it to dmac
+	 * so redirecting to ingress side of the same interface doesn't fail.
+	 */
+	if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
+		return DROP_INVALID;
+	if (eth_store_daddr(ctx, &dst_mac, 0) != 0)
+		return DROP_WRITE_ERROR;
```

Comment:
Did you check this is needed? It might be but I'm also wondering if we're not doing these rewrites too often.

Reply:
Definitely needed. The ingress packet path will drop the redirect if the MAC does not correspond to the receiving device. It occurs in https://elixir.bootlin.com/linux/latest/source/net/ethernet/eth.c#L155